src/lib/jsonParseV2/parseRegistration.cpp

```diff
@@ -0,0 +1,370 @@
+    //
+    if      (supportedForwardingMode == "none")   providerP->supportedForwardingMode = ngsiv2::ForwardNone;
+    else if (supportedForwardingMode == "query")  providerP->supportedForwardingMode = ngsiv2::ForwardQuery;
+    else if (supportedForwardingMode == "update") providerP->supportedForwardingMode = ngsiv2::ForwardUpdate;
+    else if (supportedForwardingMode == "all")    providerP->supportedForwardingMode = ngsiv2::ForwardAll;
+    else
+    {
+      *errorStringP = "invalid value of /supportedForwardingMode/";
+      return false;
+    }
+  }
+  else
+  {
+    providerP->supportedForwardingMode = ngsiv2::ForwardAll;
+  }
+
+  return true;
+}
+
+
+
+/* ****************************************************************************
+*
+* parseRegistration -
+*/
+std::string parseRegistration(ConnectionInfo* ciP, ngsiv2::Registration* regP)
+{
+  std::string          errorString;
+  rapidjson::Document  document;
+
+  document.Parse(ciP->payload);
+
+  if (document.HasParseError())
+  {
+    return badInput(ciP, ERROR_DESC_PARSE);
+  }
+
+  if (!document.IsObject())
+  {
+    return badInput(ciP, ERROR_DESC_BAD_REQUEST_NOT_A_JSON_OBJECT);
+  }
+
+  if (document.ObjectEmpty())
+  {
+    return badInput(ciP, ERROR_DESC_BAD_REQUEST_EMPTY_PAYLOAD);
+  }
+
+
+  //
+  // Extracting field "description"
+  //
+  Opt<std::string> description = getStringOpt(document, "description");
+
+  if (!description.ok())
+  {
+    return badInput(ciP, "invalid description");
+  }
+  else if (description.given)
+  {
+    regP->description = description.value;
+    if (regP->description.length() > MAX_DESCRIPTION_LENGTH)
+    {
+      return badInput(ciP, "max description length exceeded");
+    }
+
+    if (forbiddenChars(regP->description.c_str()))
+    {
+      return badInput(ciP, "forbidden characters in description");
+    }
+  }
+
+
+  //
+  // Extracting field "dataProvided"
+  //
+  if (!document.HasMember("dataProvided"))
+  {
+    return badInput(ciP, "the field /dataProvided/ is missing in payload");
+  }
+
+  const rapidjson::Value& dataProvided = document["dataProvided"];
+
+  if (dataProvidedParse(ciP, &regP->dataProvided, dataProvided, &errorString) == false)
+  {
+    return badInput(ciP, errorString);
+  }
+
+
+  //
+  // Extracting field "provider"
+  //
+  if (!document.HasMember("provider"))
+  {
+    return badInput(ciP, "the field /provider/ is missing in payload");
+  }
+
+  const rapidjson::Value& provider = document["provider"];
+
+  if (providerParse(ciP, &regP->provider, provider, &errorString) == false)
+  {
+    return badInput(ciP, errorString);
+  }
+
+
+  //
+  // Extracting field "expires"
+  //
+  if (document.HasMember("expires"))
+  {
+    const rapidjson::Value& expires = document["expires"];
+
+    if (!expires.IsString())
+    {
+      return badInput(ciP, "/expires/ must be a JSON string");
+    }
+
+    regP->expires = parse8601Time(expires.GetString());
+  }
+  else
+  {
+    regP->expires = 0x7FFFFFFFFFFFFFFF;
```

Comment:
Setting expire to MAXLONGLONG is the same strategy we are using for permanent subscriptions?

Reply:
No idea.
I tried first with -1, didn't work, then I tried with "MAX UNSIGNED LONG LONG". It worked

Reply:
This is how this is done in subscription case (from parseSubscriptions.cpp):

```
  // Expires field
  Opt<std::string> expiresOpt = getStringOpt(document, "expires");

  if (!expiresOpt.ok())
  {
    return badInput(ciP, expiresOpt.error);
  }
  else if (expiresOpt.given)
  {
    std::string  expires = expiresOpt.value;
    int64_t      eT      = -1;

    if (expires.empty())
    {
        eT = PERMANENT_SUBS_DATETIME;
    }
    else
    {
      eT = parse8601Time(expires);
      if (eT == -1)
      {
        return badInput(ciP, "expires has an invalid format");
      }
    }

    subsP->expiresProvided = true;
    subsP->expires = eT;
  }
  else if (!update)
  {
    subsP->expires = PERMANENT_SUBS_DATETIME;
  }
```

Looking to PERMANENT_SUBS_DATETIME at the end the solution seems to be similar, but for the sake of homogeneity, I'd suggest to have both versions of the code closer as possible. The PERMANENT_SUBS_DATETIME name could be done more general to be used in both places (e.g. PERMANENT_EXPIRES_DATETIME).

(the `if (!update)` is probably not relevant at the present moment, but it could be useful when PATCH get implemented)

Reply:
I can't see any usage of PERMANENT_SUBS_DATETIME anywhere except setting it and using it to avoid to render 'expires'.  Seems to me it should be present somewhere where the subscription is checked for validity.

The value of PERMANENT_SUBS_DATETIME is set to a randow value of 9e18 or something like that.
Pretty bad solution, all of this.  Would be so easy to just set it to MAX_LONG_LONG (0x7FFFFFFFFFFFFFFF) and forget about it ...  FIXME added in `globals.h` about this.

Reply:
> I can't see any usage of PERMANENT_SUBS_DATETIME anywhere except setting it and using it to avoid to render 'expires'. Seems to me it should be present somewhere where the subscription is checked for validity.

That's is normal. As far as I remember, the check is done at DB query time based on current time.

> The value of PERMANENT_SUBS_DATETIME is set to a randow value of 9e18 or something like that. Pretty bad solution, all of this. Would be so easy to just set it to MAX_LONG_LONG (0x7FFFFFFFFFFFFFFF) and forget about it ... FIXME added in globals.h about this.

Not so random, but 9e18 is pretty large (according to https://www.epochconverter.com/ is a date in March 2255 :). I like the PERMANENT_SUBS_DATETIME name for the constan, it refers to the purpose more than to the content. I also agree to enlarge its value to MAX_LONG_LONG. What about the following?

```
# define   PERMANENT_SUBS_DATETIME   MAX_LONG_LONG
```

Reply:
Yes, 
```
#define PERMANENT_EXPIRES_DATETIME 0x7FFF.....
```
is what I am using right now. 
It makes zero sense to pick a random value in the middle (it could fail - 30 years from now!!! :-D). That's not the reason. It is simply more logical to pick -1 as 'forever'.

We could also pick the value of zero as *indefinite*, but then we'd have to check for **0** when checking the expiration date. That makes the broker somewhat (almost nothing) slower. -1 is the better choice

Reply:
It is not random, it is 200 years in the future. This was checked when it was introduced. We cannot use -1 or 0, as the value is introduced in the DB and when MongoDB evaluates the query (somethign like `db.csubs.find({... , expires: {$gt: now} })` -1 or will always fail the check.

Just to now, to which date equals 0x7FFF.... ?

Reply:
200 years in the future, more or less ... seems pretty random to me :-)
More logical would be to use one of the extremes.
Not very important, 64 bits are many many bits for counting seconds ... 0x7FFF ... if 32 bits can count from 1970 to 2038 (I think it is): 68 years ... Well, 64 bits can count 4000 millions times that => ending up in the year ... 280000 million years ... :-D

Ridiculous :-)
However, I still believe 0x7FFF is a better choice than 'around 200 years from now'

Reply:
Values in the middle have the advantage of avoiding unwanted "edge effects", e.g. summing up something close to the last value will overflow to negative and things like that.

Having said that, I don't think is the case for expiration storage as they as far as I remember don't sum or any other operation in our logic, so changing to 0x7FFF is fine to me.

Reply:
Fixed in 1a49367